tracing: first tracepoints and documentation on User-Space, Statically Defined Tracing (USDT)

0xB10C opened the pull request:
This PR adds documentation for User-Space, Statically Defined Tracing (USDT) as well as three tracepoints (including documentation and usage examples).

## Context
The `TRACEx` macros for tracepoints and build system changes for USDT were merged in https://github.com/bitcoin/bitcoin/pull/19866 earlier this year. Issue https://github.com/bitcoin/bitcoin/issues/20981 contains discussion about potential tracepoints and guidelines for adding them (also documented with this PR). USDT was a topic in a [core-dev-meeting discussion](https://bitcoin.jonasschnelli.ch/ircmeetings/logs/bitcoin-core-dev/2021/bitcoin-core-dev.2021-01-21-19.00.moin.txt) on 21st Jan, 2021.

- [collabora.com: An eBPF overview, part 1: Introduction](https://www.collabora.com/news-and-blog/blog/2019/04/05/an-ebpf-overview-part-1-introduction/)
- [collabora.com: An eBPF overview, part 2: Machine & bytecode](https://www.collabora.com/news-and-blog/blog/2019/04/15/an-ebpf-overview-part-2-machine-and-bytecode/)
- [Brendan D. Gregg's blog posts, and book on on eBPF](http://www.brendangregg.com/)
- [Brendan D. Gregg: Linux bcc/BPF Node.js USDT Tracing](http://www.brendangregg.com/blog/2016-10-12/linux-bcc-nodejs-usdt.html)


## USDT? Stablecoin?

User-space, Statically Defined Tracing (USDT) allows for more observability during development, debugging, code review, and production usage. The tracepoints make it possible to keep track of custom statistics and enable detailed monitoring of otherwise hidden internals and have little to no performance impact when unused. Linux kernels (4.x or newer) can hook into the tracepoints and execute [eBPF] programs in a kernel VM once the tracepoint is called.

This PR includes, for example, tracepoints for in- and outbound P2P messages.

```
USDT and eBPF Overview
======================

                ┌──────────────────┐            ┌──────────────┐
                │ tracing script   │            │ bitcoind     │
                │==================│      2.    │==============│
                │  eBPF  │ tracing │      hooks │              │
                │  code  │ logic   │      into┌─┤►tracepoint 1─┼───┐ 3.
                └────┬───┴──▲──────┘          ├─┤►tracepoint 2 │   │ pass args
            1.       │      │ 4.              │ │ ...          │   │ to eBPF
    User    compiles │      │ pass data to    │ └──────────────┘   │ program
    Space    & loads │      │ tracing script  │                    │
    ─────────────────┼──────┼─────────────────┼────────────────────┼───
    Kernel           │      │                 │                    │
    Space       ┌──┬─▼──────┴─────────────────┴────────────┐       │
                │  │  eBPF program                         │◄──────┘
                │  └───────────────────────────────────────┤
                │ eBPF kernel Virtual Machine (sandboxed)  │
                └──────────────────────────────────────────┘

1. The tracing script compiles the eBPF code and loads the eBFP program into a kernel VM
2. The eBPF program hooks into one or more tracepoints
3. When the tracepoint is called, the arguments are passed to the eBPF program
4. The eBPF program processes the arguments and returns data to the tracing script
```

The two main [eBPF] front-ends with support for USDT are [bpftrace] an [BPF Compiler Collection (BCC)]. BCC is used for complex tools and daemons and `bpftrace` is preferred for one-liners and shorter scripts. Example tracing scripts for both are provided with this PR.

[eBPF]: https://ebpf.io/
[bpftrace]: https://github.com/iovisor/bpftrace
[BPF Compiler Collection (BCC)]: https://github.com/iovisor/bcc

This PR adds three tracepoints:
- `net:inbound_message`
- `net:outbound_message`
- `valildation:block_connected`

See `doc/tracing.md` and `contrib/tracing/` for documentation and example tracing scripts. 

## Open Questions (Not in scope for this PR)
-  How to use these tracepoints under macOS?
-  Release builds with USDT support?
-  Should and can the tracepoints be automatically tested?

## Todo (before undraft)
- [x] bcc example showing how to read raw P2P messages up to 32kb
- [x] document that you need `sys/sdt.h` from `systemtap` for USDT support in Bitcoin Core (`apt install systemtap-sdt-dev` on debian-like). See https://github.com/bitcoin/bitcoin/pull/19866/commits/933ab8a720cb9b3341adec4109cffb6dc5b322a5
- [x] release notes?

0xB10C committed:
doc: Add initial USDT documentation

Both added files are extended in the following commits.

doc/usdt.md is based on earlier work by laanwj.

Co-authored-by: W. J. van der Laan <laanwj@protonmail.com>

0xB10C committed:
tracing: Tracepoints for in- and outbound P2P msgs

Can be used to monitor in- and outbound node traffic.

Based on ealier work by jb55.

Co-authored-by: William Casarin <jb55@jb55.com>

0xB10C committed:
tracing: Tracepoint for connected blocks

Can, for example, be used to benchmark block connections.

adamjonas commented on contrib/tracing/README.md:
````diff
@@ -0,0 +1,206 @@
+…
+```
+
+Showing recent P2P messages between our onde and their node for a selected peer.
````
```suggestion
Showing recent P2P messages between our node and their node for a selected peer.
```

jonatack replied:
perhaps just omit "their node for "

0xB10C replied:
done both, thanks!

practicalswift commented:
Concept ACK

Excellent work! :)

adamjonas commented:
Concept ACK. Nice work!

jonatack commented:
Concept ACK!

jonatack commented on contrib/tracing/README.md:
````diff
@@ -43,3 +43,88 @@ example, to point to release builds if needed. See the
+```
+    ----------------------------------------------------------------------
+    |                PEER 16 (4faketorv2pbfu7x.onion:8333)               |
+    | OUR NODE              outbound-full-relay               THEIR NODE |
````
s/THEIR NODE/PEER/

0xB10C replied:
done, thanks!

jonatack commented on contrib/tracing/README.md:
```diff
@@ -43,3 +43,88 @@ example, to point to release builds if needed. See the
+### log_p2p_traffic.bt
+
+A bpftrace script logging information about inbound and outbound P2P network
+messages. Based on the `net:inbound_message` and `net:outbound_message`
```
Wishlist: this for disconnection and eviction events, with the message, connection direction and type, peer id and address.

0xB10C replied:
Agree! We discussed something similar in https://github.com/bitcoin/bitcoin/issues/20981#issuecomment-766096054 too. Thanks for the input on the arguments! I'll resolve this in favor of https://github.com/bitcoin/bitcoin/issues/20981#issuecomment-845833323 :)

jonatack replied:
(could add the network the peer connected through to the args list)

0xB10C replied:
@jonatack  there's https://github.com/bitcoin/bitcoin/pull/25832 now!

jnewbery commented on src/net.cpp:
```diff
@@ -2959,11 +2960,21 @@ bool CConnman::NodeFullyConnected(const CNode* pnode)
 {
     size_t nMessageSize = msg.data.size();
-    LogPrint(BCLog::NET, "sending %s (%d bytes) peer=%d\n",  SanitizeString(msg.m_type), nMessageSize, pnode->GetId());
+    auto sanitizedType = SanitizeString(msg.m_type);
```
Do we need to sanitize this? the message types for all outgoing messages are hardcoded in our binary so should all be safe.

0xB10C replied:
Agree, will remove it from the tracepoint and logging.

jnewbery reviewed (commented):
Concept ACK. Excellent work, @0xB10C!

0xB10C commented on doc/tracing.md:
```diff
@@ -0,0 +1,263 @@
+                │ eBPF kernel Virtual Machine (sandboxed)  │
+                └──────────────────────────────────────────┘
+
+1. The tracing script complies the eBPF code and loads the eBFP program into a kernel VM
```
```suggestion
1. The tracing script compiles the eBPF code and loads the eBFP program into a kernel VM
```

0xB10C commented:
We'll cover this PR in the PR Review Club this Wednesday (May, 26th).

https://bitcoincore.reviews/22006

0xB10C commented:
I added an example (`log_raw_p2p_msgs.py`) showcasing and documenting some USDT and eBPF limitations. Tracing P2P messages was one of the motivations for USDT in https://github.com/bitcoin/bitcoin/pull/19866 as an potential alternative to https://github.com/bitcoin/bitcoin/pull/19509. Some P2P messages like `tx` and `block` can be larger than the maximum allocation size of 32kb in eBPF. This means, we can't submit the complete P2P message to tracing scripts from the eBPF VM. Smaller messages aren't a problem. Larger messages are cut off after about 32kb in the `log_raw_p2p_msgs.py` example and a warning is printed.

Something to be aware of when adding tracepoints and developing tracing scripts.

laanwj commented:
Impressive work! Thanks for getting back to this.

>  How to use these tracepoints under macOS?

FWIW I think it's ok to keep this out of the scope of this PR. Can extend this later. Let's focus on Linux first.

> Release builds with USDT support?

Same for this. I think it's a no-brainier to have these probes in the release builds eventually, but no need to do so here. Mind that currently we have a duplicate build system (guix and gitian) and the focus for 0.22 (which is not that far away!) is ironing out issues for the guix build, as well as trying to do a guix build for the release. This is a temporary situation that complicates adding anything.

jb55 commented on src/validation.cpp:
```diff
@@ -1975,6 +1976,16 @@ bool CChainState::ConnectBlock(const CBlock& block, BlockValidationState& state,
+        block.vtx.size(),
+        nInputs,
+        nSigOpsCost,
+        GetTimeMicros() - nTimeStart, // in microseconds (µs)
```
although this is convenient, it does use up a valuable argument slot. we could achieve the same thing with a uprobe:

```bpftrace
uprobe:bitcoind:"CChainState::ConnectBlock(CBlock const&, BlockValidationState&, CBlockIndex*, CCoinsViewCache&, CChainParams const&, bool)" 
{ 
    @connect_start = nsecs;
}

usdt:bitcoind:validation:block_connected
{
    @t = nsecs - @connect_start;
}
```

(untested)

alternatively if we want a more stable interface we could just stick a usdt at the start as well.

0xB10C replied:
I'm not sure if we want to add two tracepoints (one fn start, one fn end) for measuring function timing. This would also require us to pass and handle some kind of identifier for functions running in parallel (not the case for `ConnectBlock`).

jb55 reviewed (commented):
Concept ACK!

jb55 commented on src/validation.cpp:
```diff
@@ -1975,6 +1976,16 @@ bool CChainState::ConnectBlock(const CBlock& block, BlockValidationState& state,
     LogPrint(BCLog::BENCH, "    - Callbacks: %.2fms [%.2fs (%.2fms/blk)]\n", MILLI * (nTime6 - nTime5), nTimeCallbacks * MICRO, nTimeCallbacks * MILLI / nBlocksTotal);
 
+    TRACE7(validation, block_connected,
+        block.GetHash().ToString().c_str(),
```
It's probably fine in this case, but we should avoid passing formatted data to USDTs when we can format it on the bpf side for performance reasons, especially in tight loops. It looks like we're already passing `GetHash().data()`, can we format this on the ebpf side? I remember having trouble doing this in bpftrace, but perhaps that's a temporary issue until bpftrace gets better buffer formatting.

0xB10C replied:
Agree, ideally we'd only pass the `GetHash().data()` as first argument.

>  It looks like we're already passing GetHash().data(), can we format this on the ebpf side? I remember having trouble doing this in bpftrace, but perhaps that's a temporary issue until bpftrace gets better buffer formatting.

Right, the current `%r` printf formatting for buffers in bpftrace isn't suitable for printing header hashes and txids. Adding a new formatting function to bfptrace similar to [this](https://github.com/iovisor/bpftrace/pull/1107/commits/043975141c341d2f64beaef4a99b5f2c0fbfed16#diff-e5786da510ae17c4e629faa5583e79a5f37636702844b066763fba287aa44317R829-R845)  as added in https://github.com/iovisor/bpftrace/pull/1107 doesn't look to hard. However, the block hashes and txids would still be in little-endian.

rajarshimaitra commented on contrib/tracing/log_raw_p2p_msgs.py:
```diff
@@ -0,0 +1,188 @@
+// Also see https://github.com/iovisor/bcc/issues/2306
+BPF_ARRAY(msg_arr, struct p2p_message, 1);
+
+// Two BPF perf buffers for pushing data (here P2P messages) to user -pace.
```
nit: `user space` at the end.

jnewbery commented on contrib/tracing/log_raw_p2p_msgs.py:
```diff
@@ -0,0 +1,188 @@
+
+# eBPF Limitations:
+#
+# Bitcoin P2P messages can be larger than 32kb (e.g tx, block, ...). The eBFP
```
```suggestion
# Bitcoin P2P messages can be larger than 32kb (e.g tx, block, ...). The eBPF
```

jonatack replied:
(there are at least 3 others s/eBFP/eBPF/ elsewhere)

s/e.g/e.g./

Emzy commented on contrib/tracing/connectblock_benchmark.bt:
```diff
@@ -0,0 +1,150 @@
+
+END
+{
+  print("\nHistogram of block connection times in milliseconds (ms).");
```
I get the error:
``` contrib/tracing/connectblock_benchmark.bt:138:3-71: ERROR: print() expects a map to be provided```
changing it from ```print``` to ```printf``` fixes it for me.

0xB10C replied:
Interesting.. Which version of bpftrace do you have installed? (`bpftrace --v`)

Emzy replied:
bpftrace v0.9.4 from Ubuntu 20.04.2 LTS

0xB10C replied:
Seems like histogram support for `print()` was added in a later version of bpftrace (I've been using `v0.11.4`). I've changed it to `printf()` which should work with `v0.9` and `v0.11`.

Emzy commented on contrib/tracing/p2p_monitor.py:
```diff
@@ -0,0 +1,250 @@
+                      row_format % (peer.id, peer.total_outbound_msgs, peer.total_outbound_bytes,
+                                    peer.total_inbound_msgs, peer.total_inbound_bytes,
+                                    peer.connection_type, peer.address),
+                      curses.A_NORMAL if i + scroll == cur_list_pos else curses.A_DIM)
```
The ```curces.A_DIM``` vs ```A_NORMAL``` is not showing up in my terminal.
I think this looks much better:
```
curses.A_REVERSE if i + scroll == cur_list_pos else curses.A_NORMAL)
```

rajarshimaitra reviewed (commented):
Overall Concept ACK.

I have tested and ran the examples.

Below are some nits and bumps I faced.

Overall I think having some standard tracing points and readily available processing scripts can be very helpful for developers.

I liked the p2p message monitor, and excited to see what other tracing usage others come up with.

jamesob commented:
Concept ACK - very cool. Will review soon.

0xB10C commented on contrib/tracing/log_raw_p2p_msgs.py:
```diff
@@ -0,0 +1,188 @@
+    // lookup() might not set the pointer `msg` to NULL. Thus, the BPF-verifier
+    // requires that we check that `msg` is a NULL pointer.
+    if (msg == NULL) {
+        bpf_trace_printk("not NULL\\n");
```
Thanks! That comment was incorrect and super confusing. I've changed it to

``` c
// lookup() does not return a NULL pointer. However, the BPF verifier
// requires an explicit check that that the `msg` pointer isn't a NULL
// pointer. See https://github.com/iovisor/bcc/issues/2595
if (msg == NULL) return 1;
```

See also https://github.com/iovisor/bcc/issues/2595

0xB10C commented on doc/tracing.md:
````diff
@@ -0,0 +1,263 @@
+[`tplist` usage demonstration]: https://github.com/iovisor/bcc/blob/master/tools/tplist_example.txt
+
+```
+$ tplist -l ./src/bitcoind -v
````
Hm, I agree that this is the case for Ubuntu. There seem to be naming differences between distros. 

I've documented that there might be binary naming differences depending on distro used and added `tplist-bpfcc` as an example for Ubuntu (Ubuntu is the most commonly used Linux distro, but I'm not sure how that relates to Bitcoin Core developers).

0xB10C commented on contrib/tracing/README.md:
````diff
@@ -0,0 +1,241 @@
+works fine).
+
+```
+$ bpftrace contrib/tracing/log_p2p_traffic.bt
````
Thanks for reporting, I'll have a look. Whats your bpftrace version (`bpftrace --v`)?

laanwj replied:
How is this supposed to find the process to attach to? Or would it log metrics for any `bitcoind` instance running on the system? (it seems so!)

0xB10C replied:
The bpftrace scripts include a relative path to the bitcoind binary in `./src/bitcoind` e.g. `usdt:./src/bitcoind:validation:block_connected`. 

This is mentioned on the [examples page](https://github.com/bitcoin/bitcoin/pull/22006/files#diff-a0161c271f6b7bf15c36d70d1c3943b2eb9032e6f9c754c00914103dcf7de29dR33-R36).

jb55 replied:
I think something more general would be `usdt:*bitcoind:net:inbound_message` and then

`bpftrace -p $(pgrep bitcoind) log_p2p_traffic.bt`

eg this uprobe:

```
sudo bpftrace -p $(pgrep bitcoind) -e '
  uprobe:*bitcoind:"PeerManager::ProcessMessages(CNode*, std::atomic<bool>&)" { printf("msg %p %d\n", arg1, *arg2) }
'
```

seems to work for me...

jb55 replied:
oh hmm I don't think you need `*bitcoind`, this also worked for me?

```
sudo bpftrace -p $(pgrep bitcoind) -e '
  uprobe:"PeerManager::ProcessMessages(CNode*, std::atomic<bool>&)" { printf("node %p %d\n", arg1, *arg2) }
'
```

0xB10C commented on contrib/tracing/README.md:
````diff
@@ -0,0 +1,241 @@
+messages.
+
+```
+$ python3 contrib/tracing/p2p_monitor.py ./src/bitcoind
````
> It would be even super awesome if I could also select each message and can get the content.

I agree, however I like to keep this example slim. The idea is to provide a showcase and a code example for USDT. Advanced USDT scripts can be collected somewhere else.

0xB10C committed:
doc: document systemtap dependency

0xB10C commented:
I'm marking this as ready for review. Added a commit documenting the `systemtap` dependency for USDT and installation instructions for Debian-like and Fedora (however, didn't have a chance to test building and hooking into the tracepoints on these distros yet).

The macOS CI seems to [complain](https://github.com/bitcoin/bitcoin/pull/22006/checks?check_run_id=2781317934) about the macros in `trace.h`. Haven't figured out what the issue is here. Would appreciate it if someone running macOS could try and build `bitcoind` with this patch.

fanquake commented:
> The macOS CI seems to complain about the macros in trace.h. Haven't figured out what the issue is here. Would appreciate it if someone running macOS could try and build bitcoind with this patch.

This should be solved by #22238.

fanquake commented:
@0xB10C if you rebase this now, the CI should be fixed.

0xB10C commented:
@fanquake thank you! rebased.

theStack commented:
Concept ACK. Very nice work, looking forward to reviewing and testing this in detail.

jb55 commented:
Tested ACK 462a980ebc970da1da6f76af6db5d173cd9cf951, amazing work and great docs

theStack commented on doc/dependencies.md:
```diff
@@ -39,6 +40,7 @@ Some dependencies are not needed in all configurations. The following are some f
 * SQLite is not needed with `--disable-wallet` or `--without-sqlite`.
 * Qt is not needed with `--without-gui`.
 * If the qrencode dependency is absent, QR support won't be added. To force an error when that happens, pass `--with-qrencode`.
+* If the systemtap dependency is absent, USDT support won't compiled in. To force an error when that happens, pass `--enable-ebpf`.
```
Is the error supposed to appear during the `./configure` run? In my case, explicitely passing `--enable-ebpf` didn't throw an error when `systemtap-sdt-dev` was still missing, only `ebpf tracing    = no` was displayed on the summary at the end. Tested on Debian bullseye/sid.

0xB10C replied:
Thanks. I've tested this and the docs doesn't seem to be correct. I think assumed this to be the case similar to the `--with-qrencode` behavior in the line above.

laanwj commented:
Tested ACK 462a980ebc970da1da6f76af6db5d173cd9cf951

0xB10C commented:
Sorry for invalidating the two ACKs. Addressed https://github.com/bitcoin/bitcoin/pull/22006#discussion_r667095099 and fixed another occurrence of https://github.com/bitcoin/bitcoin/pull/22006#discussion_r646471166 I previously forgot (and rebased on top of latest master).

`git range-diff 462a980eb...8f37f5c2a`

jb55 commented:
range-diff LGTM!

ACK 8f37f5c2a562c38c83fc40234ade9c301fc4e685

MarcoFalke commented on src/net.cpp:
```diff
@@ -3017,11 +3018,20 @@ bool CConnman::NodeFullyConnected(const CNode* pnode)
 {
     size_t nMessageSize = msg.data.size();
-    LogPrint(BCLog::NET, "sending %s (%d bytes) peer=%d\n",  SanitizeString(msg.m_type), nMessageSize, pnode->GetId());
+    LogPrint(BCLog::NET, "sending %s (%d bytes) peer=%d\n", msg.m_type, nMessageSize, pnode->GetId());
```
why is this refactor(?) mixed in a non-refactor commit without any explanation?

jnewbery replied:
See https://github.com/bitcoin/bitcoin/pull/22006#discussion_r636805329

MarcoFalke replied:
It would have been good to split the refactor from the feature commit or at least mention it in the commit body. Otherwise reviewing locally is impossible and one needs to rely on GitHub, which showed in the past to be corrupting both code and metadata or be unreachable.

0xB10C replied:
Didn't consider that this needs a explanation in the commit messag.

0xB10C replied:
> It would have been good to split the refactor from the feature commit or at least mention it in the commit body. Otherwise reviewing locally is impossible and one needs to rely on GitHub, which showed in the past to be corrupting both code and metadata or be unreachable.

Yes, makes sense.

rob-scheepens commented:
As BPF CO-RE arguably is the future direction for BPF, it's perhaps interesting to explore using libbpf/CO-RE instead of BCC. See https://nakryiko.com/posts/bpf-portability-and-co-re/#bpf-co-re-as-of-2021 and https://nakryiko.com/posts/bcc-to-libbpf-howto-guide/#bpf-code-conversion.

jb55 commented:
On Fri, Mar 18, 2022 at 07:27:11AM -0700, rob-scheepens wrote:
>As BPF CO-RE arguably is the future direction for BPF, it's perhaps interesting to explore using libbpf/CO-RE instead of BCC. See https://nakryiko.com/posts/bpf-portability-and-co-re/#bpf-co-re-as-of-2021 and https://nakryiko.com/posts/bcc-to-libbpf-howto-guide/#bpf-code-conversion.

This looks great! It might be a fun exercise to convert some of our bcc tools to this for anyone wanting to get more into bitcoin tracing.

0xB10C commented:
@rob-scheepens agree that BPF CO-RE and libbpf is the way to go for future tracing programs. I've looked into this a while ago, but it seems libbpf doesn't support USDT tracepoints yet (see https://github.com/iovisor/bcc/issues/3007). However, there was a mention of a soon to be open sourced libusdt library developed and used in production at Meta by @anakryiko [here](https://lore.kernel.org/bpf/CAEf4BzYRLxzVHw00DUphqqdv2m_AU7Mu=S0JF0PZYN40hBvHgA@mail.gmail.com/). Looking forward to testing this once available!

anakryiko commented:
I'm actually cleaning up the code a bit and preparing patches for submission to include proper USDT support in libbpf itself. So be on the lookout for upstream patches this week (or at worst next week, if I get distracted with other stuff this week).

anakryiko commented:
> I'm actually cleaning up the code a bit and preparing patches for submission to include proper USDT support in libbpf itself. So be on the lookout for upstream patches this week (or at worst next week, if I get distracted with other stuff this week).

BTW, USDT support was added into libbpf a while ago. See bpf_program__attach_usdt() API and SEC("usdt") examples in kernel selftests/bpf. libbpf-bootstrap also has USDT example (https://github.com/libbpf/libbpf-bootstrap/blob/master/examples/c/usdt.c and https://github.com/libbpf/libbpf-bootstrap/blob/master/examples/c/usdt.bpf.c).

fanquake commented:
This change was part of 23.0, and had a release note, so removing "Needs release note".
